UnwrapBlog.md

```diff
@@ -0,0 +1,92 @@
+new Vpc({ ipAddresses: computeAnArrayAnyOfTheAboveAsynchrousnously() }); // etc. etc.
+```
+
+It's super flexible, and let's people naturally produce the values in whatever way they want without having to work too hard to make it conform to some very narrow type.
+
+However, while easy to produce, this can sometimes be difficult to consume.  For example, if you wanted to take in that `Input<Input<string>[]>` and then check if there was a very specific value in it, you might have to go through and do a bunch of work.  You'd have to first deal with potentially getting a `Input<string>[]` or a `Promise<Input<string>[]>` or even an `Output<Input<string>[]>`. Then, once you even got to the underlying array, you'd have to deal with each potential element in it being a `string`... or a `Promise<string>`... or a... well... you get the idea.  It's not fun 😃
+
+Up till now, we'd supplied some simple helpers to make this more managable.  Our helpers would effectively 'unwrap' one layer of this sort of structure allowing you take the external value and deal with an internal value of a known shape.  So it effectively worked like this:
+
+```ts
+function unwrap<T>(input: Input<T>): T { /* ... */ }
+
+// so, you could then write:
+const ipAddresses: Input<string>[] = unwrap(vpc.ipAddresses);
+```
+
+This helped things somewhat.  But it still meant that as you processed that array you had to then 
+unwrap each and every element.  As our data model has grown, and we've tackled more and more
+complex domains, we started seeing code start to form the programming [Pyramid of Doom](https://en.wikipedia.org/wiki/Pyramid_of_doom_(programming)), where you just kept on going deeper and deeper, unwrapping all the way.  We started seeing this happen so often that we wanted to see if we there was anything we could do to make life much nicer for people using our APIs and using these helpers.
+
+Our core goal was to change our 'unwrap' function from just peeling away one layer of this onion, to peeling away *all* layers.  So, for example, if you started with an `Input<Input<string>[]>` you would just get back a plain old `string[]` that you could use completely naturally in all the ways you'd expect.  We were initially uncertain if we could make this work, but thanks to the power of JavaScript and the expressivity of TypeScript, we think we found a great solution here.
+
+We knew that providing the implementation such a function itself was not going to the true hard part.  After all, we already had the code to peel back one layer.  So peeling back all the layers was more a matter of just having the function call itself recursively when encountering things like arrays (or other objects that themselves contained `Input`s).  The hard part came in when were trying figure out how would we even express this in TypeScript?.  After all, what's the type signature when you effectively to say:
```

Comment:
Making the point that this function is in fact really *easy* to write in JavaScript, but just not as easy to *type* might help here.  That's sort of the crux of why this is interesting/exciting - a reasonably easy/common thing to do in JavaScript, which can now be typed accurately in TypeScript.

Reply:
I thought that was made with: "The hard part came in when were trying figure out how would we even express this in TypeScript?"

Reply:
Yep - point is made - was more thinking of emphasizing how easy/normal it is to write this sort of thing in JavaScript - and hence why this is a problem/solution that is broadly interesting.

Reply:
I've tried to clean this up a tad :)